rskj-core/src/main/java/co/rsk/util/ContractUtil.java

```diff
@@ -0,0 +1,125 @@
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+package co.rsk.util;
+
+import co.rsk.PropertyGetter;
+import co.rsk.core.Coin;
+import co.rsk.core.RskAddress;
+import co.rsk.core.bc.AccountInformationProvider;
+import org.ethereum.config.Constants;
+import org.ethereum.core.CallTransaction;
+import org.ethereum.core.SignatureCache;
+import org.ethereum.core.Transaction;
+import org.ethereum.crypto.HashUtil;
+import org.ethereum.solidity.SolidityType;
+import org.ethereum.util.ByteUtil;
+import org.ethereum.vm.DataWord;
+
+import java.math.BigInteger;
+import java.util.Arrays;
+
+public class ContractUtil {
+    private static final String SWAPS_MAP_POSITION = "0000000000000000000000000000000000000000000000000000000000000000";
+
+    private ContractUtil() {
+        throw new IllegalStateException("Utility class");
+    }
+
+    public static byte[] calculateSwapHash(Transaction newTx, SignatureCache signatureCache) {
+
+        CallTransaction.Invocation invocation = getTxInvocation(newTx);
+
+        // Get arguments from the invocation object
+        byte[] preimage = (byte[]) invocation.args[0];
+        byte[] preimageHash = HashUtil.sha256(encodePacked(preimage));
+        BigInteger amount = (BigInteger) invocation.args[1];
+        DataWord refundAddress = (DataWord) invocation.args[2];
+        BigInteger timeLock = (BigInteger) invocation.args[3];
+
+
+        // Remove the leading zeroes from the arguments bytes and merge them
+        byte[] argumentsBytes = encodePacked(
+                preimageHash,
+                amount,
+                newTx.getSender(signatureCache),
+                new RskAddress(ByteUtil.toHexString(refundAddress.getLast20Bytes())),
+                timeLock
+        );
+
+        return HashUtil.keccak256(argumentsBytes);
+    }
+
+    public static byte[] encodePacked(Object...arguments) {
+        byte[] encodedArguments = new byte[]{};
+
+        for(Object arg: arguments) {
+            byte[] encodedArg = new byte[]{};
+            if(arg instanceof byte[]) {
+                SolidityType bytes32Type = SolidityType.getType(SolidityType.BYTES32);
+                encodedArg = bytes32Type.encode(arg);
+            } else if(arg instanceof RskAddress) {
+                SolidityType addressType = SolidityType.getType(SolidityType.ADDRESS);
+                byte[] encodedAddress = addressType.encode(((RskAddress) arg).toHexString());
+                encodedArg = org.bouncycastle.util.Arrays.copyOfRange(encodedAddress, 12, encodedAddress.length);
+            } else if(arg instanceof BigInteger) {
+                SolidityType uint256Type = SolidityType.getType(SolidityType.UINT);
+                encodedArg = uint256Type.encode(arg);
+            }
+
+            encodedArguments = ByteUtil.merge(encodedArguments, encodedArg);
+        }
+
+        return encodedArguments;
+    }
```

Comment:
Some suggestion with a simplification.

```suggestion
    public static byte[] encodePacked(Object...arguments) {
        return Arrays.stream(arguments)
                .map(ContractUtil::encodeArgument)
                .reduce(ByteUtil::merge)
                .orElse(new byte[]{});
    }

    private static byte[] encodeArgument(Object arg) {
        if(arg instanceof byte[]) {
            return SolidityType.getType(SolidityType.BYTES32).encode(arg);
        } else if(arg instanceof RskAddress) {
            byte[] encodedAddress = SolidityType.getType(SolidityType.ADDRESS).encode(((RskAddress) arg).toHexString());
            return org.bouncycastle.util.Arrays.copyOfRange(encodedAddress, 12, encodedAddress.length);
        } else if(arg instanceof BigInteger) {
            return SolidityType.getType(SolidityType.UINT).encode(arg);
        }
        return new byte[]{};
    }
```

Reply:
I think it's debatable, for big arrays the stream sometimes wins on performance. I prefer because makes the code way more readable. But you could enclose the if-else chain in a method with a meaningful name stating your intention. This would help in the readability of this if-else chain. Something like:

```
private static byte[] encodeArgumentAccordingInstanceType(Object arg) {
        if(arg instanceof byte[]) {
            return SolidityType.getType(SolidityType.BYTES32).encode(arg);
        } else if(arg instanceof RskAddress) {
            byte[] encodedAddress = SolidityType.getType(SolidityType.ADDRESS).encode(((RskAddress) arg).toHexString());
            return org.bouncycastle.util.Arrays.copyOfRange(encodedAddress, 12, encodedAddress.length);
        } else if(arg instanceof BigInteger) {
            return SolidityType.getType(SolidityType.UINT).encode(arg);
        }
        return new byte[]{};
}
```